src/harness/reference_models/ppa/ppa.py

```diff
@@ -0,0 +1,150 @@
+                                for lat, lon, gain in zip(latitudes, longitudes, antenna_gain)])
+  # Generating lat, lon for Contour
+  contour_lats, contour_lons, _ = zip(*[vincenty.GeodesicPoint(install_param['latitude'],
+                                                               install_param['longitude'], cn, az)
+                                        for cn, az in zip(contour_pts, azimuths)])
+  return geometry.Polygon(zip(contour_lons, contour_lats)).buffer(0)
+
+
+def _ClipPpaByCensusTract(contour_union, pal_records):
+  # Get the Census Tract for Each Pal Record and Convert it to Shapely Geometry
+  census_tracts_for_pal = [geometry.shape(census_tract_driver.GetCensusTract(pal['license']
+                                                                             ['licenseAreaIdentifier'])
+                                          ['features'][0]['geometry']).buffer(0) for pal in pal_records]
+  census_tracts_union = ops.cascaded_union(census_tracts_for_pal)
+  return contour_union.intersection(census_tracts_union)
+
+
+def _ConvertToGeoJson(ppa_polygon):
+  if ppa_polygon.type == 'MultiPolygon':
+    ppa_geojson = geojson.FeatureCollection(
+      [geojson.Feature(geometry=[polygon for polygon in ppa_polygon])])
+  else:
+    ppa_geojson = geojson.FeatureCollection(
+      [geojson.Feature(geometry=ppa_polygon)])
+  return ppa_geojson
+
+
+def ConfigureCensusTractDriver(census_tract_dir=None):
+  """Configure the Census Tract driver.
+
+  Inputs:
+    census_tract_dir: if specified, changes the census tract default directory.
+  """
+  if census_tract_dir is not None:
+    census_tract_driver.SetCensusTractDirectory(census_tract_dir)
+
+
+def ConfigureNlcdDriver(nlcd_dir=None, cache_size=None):
+  """Configure the NLCD driver.
+
+  Inputs:
+    nlcd_dir: if specified, changes the NLCD Data default directory.
+    cache_size:  if specified, change the NLCD tile cache size.
+  Note: The memory usage is about cache_size * 50MB.
+
+  """
+  if nlcd_dir is not None:
+    nlcd_driver.SetTerrainDirectory(nlcd_dir)
+  if cache_size is not None:
+    nlcd_driver.SetCacheSize(cache_size)
+
+
+def PpaCreationModel(devices, pal_records):
+  """Create a PPA Polygon based on the PAL Records and Device Information
+  Args:
+    devices: (List) A list containing device information.
+    pal_records: (List) A list containing pal records.
+  Returns:
+     A GeoJSON Dictionary of PPA Polygon
+  """
+  # TODO: Add Validation for Inputs
+  pool = futures.ProcessPoolExecutor(multiprocessing.cpu_count())
+  # Create Contour for each CBSD
+  device_polygon = list(pool.map(_GetPolygon, devices))
+  # Create Union of all the CBSD Contours and Check for hole
+  # after Census Tract Clipping
+  contour_union = ops.cascaded_union(device_polygon)
+  ppa_polygon = _ClipPpaByCensusTract(contour_union, pal_records)
+  return _ConvertToGeoJson(utils.PolyWithoutSmallHoles(ppa_polygon))
```

Comment:
Isn't there also a smoothing function that is applied (before? after?) the holes are removed?

Reply:
I think this is taken care of:  smoothing by the Hamming filter applied on each individual device, and removal of the small holes of the final contour.